inference-engine/src/mkldnn_plugin/nodes/mkldnn_matrix_nms_node.cpp

```diff
@@ -0,0 +1,512 @@
+        outDataConf.emplace_back(TensorDescCreatorTypes::ncsp, outPrecision);
+    }
+
+    addSupportedPrimDesc(inDataConf, outDataConf, impl_desc_type::ref_any);
+}
+
+bool MKLDNNMatrixNmsNode::created() const {
+    return getType() == MatrixNms;
+}
+
+namespace {
+template<typename T, bool gaussian>
+struct decay_score;
+
+template<typename T>
+struct decay_score<T, true> {
+    T operator()(T iou, T max_iou, T sigma) {
+        return std::exp((max_iou * max_iou - iou * iou) * sigma);
+    }
+};
+
+template<typename T>
+struct decay_score<T, false> {
+    T operator()(T iou, T max_iou, T sigma) { return (1. - iou) / (1. - max_iou + 1e-10f); }
+};
+
+template<class T>
+static inline T BBoxArea(const T *box, const bool normalized) {
+    if (box[2] < box[0] || box[3] < box[1]) {
+        // If coordinate values are is invalid
+        // (e.g. xmax < xmin or ymax < ymin), return 0.
+        return static_cast<T>(0.);
+    } else {
+        const T w = box[2] - box[0];
+        const T h = box[3] - box[1];
+        if (normalized) {
+            return w * h;
+        } else {
+            // If coordinate values are not within range [0, 1].
+            return (w + 1) * (h + 1);
+        }
+    }
+}
+
+template<class T>
+static inline T
+intersectionOverUnion(const T *box1, const T *box2, const bool normalized) {
+    if (box2[0] > box1[2] || box2[2] < box1[0] || box2[1] > box1[3] ||
+        box2[3] < box1[1]) {
+        return static_cast<T>(0.);
+    } else {
+        const T inter_xmin = std::max(box1[0], box2[0]);
+        const T inter_ymin = std::max(box1[1], box2[1]);
+        const T inter_xmax = std::min(box1[2], box2[2]);
+        const T inter_ymax = std::min(box1[3], box2[3]);
+        T norm = normalized ? static_cast<T>(0.) : static_cast<T>(1.);
+        T inter_w = inter_xmax - inter_xmin + norm;
+        T inter_h = inter_ymax - inter_ymin + norm;
+        const T inter_area = inter_w * inter_h;
+        const T bbox1_area = BBoxArea<T>(box1, normalized);
+        const T bbox2_area = BBoxArea<T>(box2, normalized);
+        return inter_area / (bbox1_area + bbox2_area - inter_area);
+    }
+}
+
+struct Rectangle {
+    Rectangle(float x_left, float y_left, float x_right, float y_right)
+            : x1{x_left}, y1{y_left}, x2{x_right}, y2{y_right} {}
+
+    Rectangle() = default;
+
+    float x1 = 0.0f;
+    float y1 = 0.0f;
+    float x2 = 0.0f;
+    float y2 = 0.0f;
+};
+
+struct BoxInfo {
+    BoxInfo(const Rectangle &r,
+            int64_t idx,
+            float sc,
+            int64_t batch_idx,
+            int64_t class_idx)
+            : box{r}, index{idx}, batch_index{batch_idx}, class_index{class_idx}, score{sc} {
+    }
+
+    BoxInfo() = default;
+
+    Rectangle box;
+    int64_t index = -1;
+    int64_t batch_index = -1;
+    int64_t class_index = -1;
+    float score = 0.0f;
+};
+
+template<typename T, bool gaussian>
+size_t nms_matrix(const T *boxes_data,
```

Comment:
looks like you can make ```nms_matrix``` as class method and don't pass attributes as function arguments
do we really need to have template parameter type for this function? if we support ```float``` only?
same for other functions

Reply:
Currently we only support fp32. It could be converted as a class method.